### PR TITLE
Implement window displaying object load failures

### LINKED ISSUE #1399: Separate window for displaying object load failures
It would be nice if we could display to the user as a OpenRCT2 window all of the objects that have failed to load and any extra info about them (exapansion, checksum failure, object type). At present set_load_objects_fail_reason is the function used for setting the single line message.

### LINKED ISSUE #5177: No notification when objects are missing
**OS:** Ubuntu 16.04
**Version:** 0.0.7
**Commit/Build:** b21378b

If you open the following park, it won't load due to missing objects (it prints that to stdout in verbose mode), but it won't display a message telling the user so.
https://github.com/OpenRCT2/OpenRCT2/files/741918/autosave_2017-01-31_23-17-53.zip

### PR DESCRIPTION
**Pull request by rwjuk**
In its current state, this partially resolves #1399, I have some more work to do (see below) that will cover more of the ideas there. This also resolves #5177.

Here's an SV6 with invalid objects for testing purposes: https://github.com/OpenRCT2/OpenRCT2/files/741918/autosave_2017-01-31_23-17-53.zip

TODO:
* [x]  Implement 'copy current item to clipboard'
* [x]  Implement 'copy whole list to clipboard'.

### CODE REVIEW

File: src/openrct2/rct2/S6Importer.cpp [hunk 1]
```diff
@@ -169,6 +169,11 @@ class S6Importer final : public IParkImporter
         return false;
     }
 
+    bool HasInvalidObjects() override
```
Comment (Gymnasiast) [inline src/openrct2/rct2/S6Importer.cpp]: You called this function  `HasInvalidObjects()`, yet it returns whether the save contains _no_ invalid objects. The functions it calls all return true on success anyway (i.e. all valid), so I'd rename this function to `HasNoInvalidObjects()` or something similar.

Comment (rwjuk) [inline src/openrct2/rct2/S6Importer.cpp]: Bugger, I knew I had it that way to start with for a reason! You're entirely right, I'll swap it back!

File: src/openrct2/rct2/S6Importer.cpp [hunk 2]
```diff
@@ -429,7 +429,7 @@ extern "C"
         {
             s6Importer->LoadSavedGame(path);
         }
-        return s6Importer->HasInvalidObjects();
+        return !s6Importer->HasNoInvalidObjects();
```
Comment (Gymnasiast) [inline src/openrct2/rct2/S6Importer.cpp]: I'm not sure if it's a good idea to have the C function act differently from the C++ counterpart. Probably better to change it as well.

Comment (rwjuk) [inline src/openrct2/rct2/S6Importer.cpp]: Yeah, good point. I think I swapped back and forth so many times on the sense of these functions that it all got a bit confused.

**Comment by IntelOrca** [comment]
The object manager should not store the invalid object entries as this data is specific to a particular load park instance. Instead loading a park should return a result object which contains a success or fail flag, and if the later, a list of objects it could not load.

**Comment by rwjuk** [comment]
@IntelOrca I'm guessing that the result object should have two flags, a success/fail and then a 'bad objects' flag, so we know whether loading the park failed for objects or some other reason?

**Comment by IntelOrca** [comment]
Well rather than a fail / success flag (that can be a helper method), you can just have an error reason enum.

File: src/openrct2/windows/object_load_error.c [hunk 3]
```diff
@@ -174,7 +174,7 @@ rct_window *window_object_load_error_open(utf8* path, object_validity_result res
 
 static void window_object_load_error_close(rct_window *w)
 {
-    SafeFree(invalid_entries);
+    //SafeFree(invalid_entries);
```
Comment (Gymnasiast) [inline src/openrct2/windows/object_load_error.c]: \*cough\*

Comment (rwjuk) [inline src/openrct2/windows/object_load_error.c]: You saw nothing 😛

File: src/openrct2/object/ObjectManager.cpp [hunk 4]
```diff
@@ -438,6 +438,55 @@ class ObjectManager final : public IObjectManager
...
     }
 
+    rct_object_entry* DuplicateObjectEntry(const rct_object_entry* original)
+    {
+        rct_object_entry * duplicate = Memory::Allocate<rct_object_entry>(sizeof(rct_object_entry));
+        duplicate->checksum = original->checksum;
+        strncpy(duplicate->name, original->name, 8);
+        
+        // Ensure the string is properly terminated, which is often not the case.
+        duplicate->name[8] = '\0';
```
Comment (janisozaur) [inline src/openrct2/object/ObjectManager.cpp]: There's no termination on these, just the id

**Comment by rwjuk** [comment]
This is (finally!) ready for review & testing. I haven't implemented all the ideas from #1399 (e.g. replacing invalid objects automatically), but this can always be added in later, and the main features that most users will value are there. It's certainly an improvement over what we have currently.

I'm also aware that the park load functions could and should be refactored to take further advantage of the `park_load_result` type, which allows specifying an error.

File: src/openrct2/object/ObjectManager.h [hunk 5]
```diff
@@ -65,12 +67,12 @@ extern "C"
 {
 #endif
 
-void *  object_manager_get_loaded_object_by_index(size_t index);
-void *  object_manager_get_loaded_object(const rct_object_entry * entry);
-uint8   object_manager_get_loaded_object_entry_index(const void * loadedObject);
-void *  object_manager_load_object(const rct_object_entry * entry);
-void    object_manager_unload_objects(const rct_object_entry * entries, size_t count);
-void    object_manager_unload_all_objects();
+void *                      object_manager_get_loaded_object_by_index(size_t index);
```
Comment (Gymnasiast) [inline src/openrct2/object/ObjectManager.h]: Why all this whitespace?

File: src/openrct2/ui/UiContext.h [hunk 6]
```diff
@@ -136,6 +136,9 @@ namespace OpenRCT2
             // HACK: This should either be implemented ourselves in libopenrct2
             //       or the mapgen height map code is moved to libopenrct2ui.
             virtual bool                ReadBMP(void * * outPixels, uint32 * outWidth, uint32 * outHeight, const std::string &path) abstract;
+
+            // Clipboard
+            virtual bool              SetClipboardText(utf8* target) abstract;
```
Comment (IntelOrca) [inline src/openrct2/ui/UiContext.h]: This should be declared as `const std::string &target` or `const utf8 * target`

File: src/openrct2/windows/object_load_error.c [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+static utf8* combine_object_names(rct_window *w)
+{
+    if (w->no_list_items > OBJECT_ENTRY_COUNT || w->no_list_items == 0) {
+        // Something's gone wrong, this shouldn't happen
+        // We don't want to allocate stupidly large amounts of memory
+        // for no reason, so bail
+        return NULL;
+    }
+    utf8* buffer;
+    utf8* line_sep = platform_get_line_separator();
```
Comment (IntelOrca) [inline src/openrct2/windows/object_load_error.c]: There is already a line separator `#define` somewhere that you can use.
Also this needs to be `freed` as your function allocates memory.

File: src/openrct2/windows/object_load_error.c [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+static void window_object_load_error_mouseup(rct_window *w, rct_widgetindex widgetIndex)
+{
+    utf8* combined_list;
+    switch (widgetIndex) {
+        case WIDX_CLOSE:
+            window_close(w);
+            break;
+        case WIDX_COPY_CURRENT:
+            if (w->selected_list_item > -1) {
+                platform_place_string_on_clipboard(strndup(invalid_entries[w->selected_list_item]->name, 8));
```
Comment (IntelOrca) [inline src/openrct2/windows/object_load_error.c]: The string you pass in needs to be freed.

File: src/openrct2/object/ObjectManager.h (hunk 5 above)
Comment (rwjuk) [inline src/openrct2/object/ObjectManager.h]: Whoa. No idea, that didn't show up locally!

File: src/openrct2/windows/object_load_error.c (hunk 8 above)
Comment (rwjuk) [inline src/openrct2/windows/object_load_error.c]: Managed to catch the leak in the case below but not this one! Will fix.

File: src/openrct2/windows/object_load_error.c (hunk 7 above)
Comment (rwjuk) [inline src/openrct2/windows/object_load_error.c]: Ahhh, it's `PLATFORM_NEWLINE`, for some reason I only searched for variations on 'separator.

**Comment by IntelOrca** [comment]
There are some issues:
* It does not show the window for SC4 files.
* It does not work with scenarios because it loads the title sequence which closes the window.

I will try and fix it and any other issues I find.

**Comment by rwjuk** [comment]
@IntelOrca As Sx4s can't have custom objects, I didn't think it was needed for those?

**Comment by IntelOrca** [comment]
Ah, but did you try deleting half of the vanilla objects?

**Comment by rwjuk** [comment]
... heh, didn't think of that! Yup, I'll hook it up for Sx4s too!

**Comment by rwjuk** [comment]
I've hooked it up for scenarios just fine - S4s is proving a little more difficult as the S4 import process is very different. May need @Gymnasiast to help me figure out how object entries are converted to their RCT2 equivalents. Are ones that don't exist skipped during conversion? It seems that way...

**Comment by Gymnasiast** [comment]
> Are ones that don't exist skipped during conversion? It seems that way...

All RCT1 objects are imported. I'm not sure how it behaves if one of the RCT2 objects it needs is missing, but since it's converted to an S6 first, I guess there shouldn't be much of a difference, really.

**Comment by IntelOrca** [comment]
There are huge amount of memory issues here, lots of memory that is not freed and potentially not freed if exceptions happen. Some memory functions not called correctly.

I will sort this out this evening by using C++ structures and RAII for as much as possible and then converting to a C struct as it goes to C code area. Its easier and safer that way.

**Comment by rwjuk** [comment]
@IntelOrca I'll make sure to look closely at the changes to see where I went wrong and how to implement things better next time.

**Comment by IntelOrca** [comment]
@rwjuk The design is okay, its just the unfortunate complication of our C -> C++ -> C code, our own confusing memory functions and general memory management being a tedious and unnecessary thing we have to do.

**Pull request merged**